test/integration_test.py

```diff
@@ -405,6 +410,54 @@ def _get_entities_by_project(self, entity_type, project_short_name):
             params['search_after_uid'] = pagination['search_after_uid']
         return entities
 
+    @unittest.skipIf(config.deployment_stage not in ('sandbox', 'nadove'),
+                     'Not polluting portal db since not in sandbox environment')
+    def test_concurrent_portal_db_crud(self):
+        """
+        Use multithreading to simulate multiple users simultaneously modifying
+        the portals database.
+        """
+
+        # Currently takes about 50 seconds and creates a 25 kb db file.
+        n_threads = 10
+        n_ops = 10
+        portal_service = PortalService()
+
+        entry_format = 'thread={};op={}'
+
+        def run(thread_count):
+            for op_count in range(n_ops):
+                mock_entry = cast(JSON, {
+                    "portal_id": "foo",
+                    "integrations": [
+                        {
+                            "integration_id": "bar",
+                            "entity_type": "project",
+                            "integration_type": "get",
+                            "entity_ids": ["baz"]
+                        }
+                    ],
+                    "mock-count": entry_format.format(thread_count, op_count)
+                })
+                portal_service._crud(lambda db: list(db) + [mock_entry])
+
+        old_db = portal_service.read()
+
+        with ThreadPoolExecutor(max_workers=n_threads) as executor:
+            for i in range(n_threads):
+                executor.submit(run, i)
```

Comment:
To collect exceptions raised in worker threads, one would need to keep track of the return values of submit() and call result() or exception() on them after all tasks have been submitted. If the tasks don't return anything the bare bones is

```
futures = [executor.submit() for i in range(…)]
assert all(f.result() is None for f in futures)
```

This would raise an exception on the first worker exception encountered and ignore any exceptions in the other futures. I think that would suffice here as long as a worker exception propagates to a test failure.